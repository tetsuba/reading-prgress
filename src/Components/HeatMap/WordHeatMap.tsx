import { HeatMapColors } from './HeatMap'

export type WordHeatMapTypes = {
    word: string
    index: number
    max: number
    color: HeatMapColors
}

export default function WordHeatMap(props: WordHeatMapTypes) {
    const rgb = 255 * (props.index / props.max)
    const numberHeatMap = {
        backgroundColor: `rgb(${255 - rgb},${255 - rgb},${255 - rgb})`
    }
    const showBadge = props.index > 1

    function getBackGroundStyles() {
        if (props.color === 'red') {
            return { backgroundColor: `rgb(255,${255 - rgb},${255 - rgb})` }
        }
        if (props.color === 'blue') {
            return {
                backgroundColor: `rgb(${255 - rgb}, ${255 - rgb / 2}, 255)`
            }
        }
        if (props.color === 'green') {
            return {
                backgroundColor: `rgb(${255 - rgb}, ${255 - rgb / 3}, ${
                    255 - rgb
                })`
            }
        }
        if (props.color === 'none') {
            return { backgroundColor: 'rgb(230,230,230)' }
        }
    }

    function getTextStyles() {
        if (props.color === 'none') {
            return 'text-gray-900'
        }
        return (props.index / props.max) * 100 > 50
            ? 'text-white'
            : 'text-gray-900'
    }

    return (
        <span
            style={getBackGroundStyles()}
            data-testid="heat-map-word"
            className={`m-6 rounded-2xl border-2 p-3`}
        >
            <span className={`font-bold ${getTextStyles()}`}>{props.word}</span>
            {showBadge && (
                <span
                    style={numberHeatMap}
                    className={`ml-4 rounded-full py-1 px-3 text-white`}
                >
                    {props.index}
                </span>
            )}
        </span>
    )
}
