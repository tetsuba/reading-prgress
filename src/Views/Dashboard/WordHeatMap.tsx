export type WordHeatMapTypes = {
    word: string
    index: number
}

function getHeatMap(index: number) {
    if (index < 2) {
        return { number: '', bg: 'border-gray-100', text: 'text-gray-900' }
    } else if (index < 5) {
        return {
            number: 'bg-red-100',
            bg: 'border-gray-100 bg-red-50',
            text: 'text-gray-900'
        }
    } else if (index < 10) {
        return {
            number: 'bg-red-300',
            bg: 'border-gray-100 bg-red-100',
            text: 'text-gray-800'
        }
    } else if (index < 15) {
        return {
            number: 'bg-red-400',
            bg: 'border-gray-200 bg-red-100',
            text: 'text-gray-700'
        }
    } else if (index < 20) {
        return {
            number: 'bg-red-500',
            bg: 'border-gray-300 bg-red-200',
            text: 'text-gray-600'
        }
    } else {
        return {
            number: 'bg-black',
            bg: 'border-gray-300 bg-red-500',
            text: 'text-white'
        }
    }
}

export default function WordHeatMap(props: WordHeatMapTypes) {
    const heatMap = getHeatMap(props.index)

    return (
        <span
            data-testid="heat-map-word"
            className={`m-6 rounded-2xl border-2 p-3 ${heatMap.bg}`}
        >
            <span className={`font-bold text-gray-900 ${heatMap.text}`}>
                {props.word}
            </span>
            <span
                className={`ml-4 rounded-full py-1 px-3 text-white ${heatMap.number}`}
            >
                {props.index}
            </span>
        </span>
    )
}
