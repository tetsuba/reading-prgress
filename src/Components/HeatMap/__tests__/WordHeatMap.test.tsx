import { render } from '@testing-library/react'
import WordHeatMap from '../WordHeatMap'

describe('WordHeatMap', () => {
    test.each(['red', 'blue', 'green', 'none'])(
        'word color is %s',
        (color: any) => {
            const { asFragment } = render(
                <WordHeatMap max={3} word="heat" index={5} color={color} />
            )
            expect(asFragment()).toMatchSnapshot()
        }
    )
})
