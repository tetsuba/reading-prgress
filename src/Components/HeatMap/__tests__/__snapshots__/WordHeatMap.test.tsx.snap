// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WordHeatMap > word color is blue 1`] = `
<DocumentFragment>
  <span
    class="m-2 rounded-2xl border-2 p-3 md:m-5"
    data-testid="heat-map-word"
    style="background-color: rgb(-170, 42.5, 255);"
  >
    <span
      class="font-bold text-white"
    >
      word
    </span>
    <span
      class="ml-1 rounded-full px-2 py-1 text-white md:ml-3"
      style="background-color: rgb(-170, -170, -170);"
    >
      5
    </span>
  </span>
</DocumentFragment>
`;

exports[`WordHeatMap > word color is green 1`] = `
<DocumentFragment>
  <span
    class="m-2 rounded-2xl border-2 p-3 md:m-5"
    data-testid="heat-map-word"
    style="background-color: rgb(-170, 113.33333333333334, -170);"
  >
    <span
      class="font-bold text-white"
    >
      word
    </span>
    <span
      class="ml-1 rounded-full px-2 py-1 text-white md:ml-3"
      style="background-color: rgb(-170, -170, -170);"
    >
      5
    </span>
  </span>
</DocumentFragment>
`;

exports[`WordHeatMap > word color is none 1`] = `
<DocumentFragment>
  <span
    class="m-2 rounded-2xl border-2 p-3 md:m-5"
    data-testid="heat-map-word"
    style="background-color: rgb(230, 230, 230);"
  >
    <span
      class="font-bold text-gray-900"
    >
      word
    </span>
    <span
      class="ml-1 rounded-full px-2 py-1 text-white md:ml-3"
      style="background-color: rgb(-170, -170, -170);"
    >
      5
    </span>
  </span>
</DocumentFragment>
`;

exports[`WordHeatMap > word color is red 1`] = `
<DocumentFragment>
  <span
    class="m-2 rounded-2xl border-2 p-3 md:m-5"
    data-testid="heat-map-word"
    style="background-color: rgb(255, -170, -170);"
  >
    <span
      class="font-bold text-white"
    >
      word
    </span>
    <span
      class="ml-1 rounded-full px-2 py-1 text-white md:ml-3"
      style="background-color: rgb(-170, -170, -170);"
    >
      5
    </span>
  </span>
</DocumentFragment>
`;
